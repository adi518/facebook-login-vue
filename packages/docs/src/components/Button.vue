<template>
  <button class="button" :class="{ 'is-icon': isIcon }" v-on="$listeners">
    <slot></slot>
  </button>
</template>

<script>
export default {
  props: {
    isIcon: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="scss" scoped>
@import '@/styles/bootstrap-required';

.button {
  border-width: 0;
  font-size: 0.825rem;
  font-family: inherit;
  padding: 0.25rem 1rem;
  border-radius: 0.25rem;
  text-transform: capitalize;
  color: rgba(139, 157, 195, 0.5);
  background-color: darken(#3c57a4, 25%);
  border: 1px solid rgba(139, 157, 195, 0.15);
  transition: color 0.5s, background-color 0.5s;

  &.is-icon {
    padding-left: 0.5rem;
    padding-right: 0.5rem;
  }

  &:hover {
    color: white;
    background-color: transparent;
    border-color: rgba(255, 255, 255, 0.25);
  }

  &:not(:first-child) {
    margin-left: 2ch;
  }

  @include media-breakpoint-down(sm) {
    display: none;
  }
}
</style>
