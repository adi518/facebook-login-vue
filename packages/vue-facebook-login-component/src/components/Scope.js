import {
  getSdk,
  login,
  logout,
  getLoginStatus,
  removeSdkScript
} from '@/helpers'

export default {
  name: 'v-facebook-login-scope',
  props: {
    value: {
      type: Object,
      default: () => ({ connected: false })
    },
    appId: {
      type: String,
      required: true
    },
    version: {
      type: String,
      default: 'v6.0' // https://developers.facebook.com/docs/graph-api/changelog/
    },
    options: {
      type: Object,
      default: () => ({})
    },
    loginOptions: {
      type: Object,
      default: () => ({ scope: 'email' })
    }
  },
  data: () => ({
    error: null,
    working: true,
    connected: false
  }),
  watch: {
    // eslint-disable-next-line no-unused-vars
    scope({ login, logout, toggleLogin, ...restScope }, _prevValue) {
      this.$emit('input', restScope)
    }
  },
  async beforeCreate() {
    await removeSdkScript()
  },
  async created() {
    const created = async () => {
      const sdk = await this.getSdk()
      if (this.error) return void 0
      this.$emit('sdk-init', { FB: sdk, scope: this.scope })
      const { status } = await getLoginStatus()
      if (status === 'connected') {
        this.connected = true
        this.$emit('login')
      }
    }
    this.async(created())
  },
  updated() {
    if (this.$slots.default?.length) {
      console.error('Slot must be scoped.')
    }
  },
  computed: {
    idle() {
      return !this.working
    },
    enabled() {
      return !this.disabled
    },
    disabled() {
      return this.working || Boolean(this.error) || !this.appId
    },
    disconnected() {
      return !this.connected
    },
    scope() {
      return {
        idle: this.idle,
        error: this.error,
        login: this.login,
        logout: this.logout,
        working: this.working,
        enabled: this.enabled,
        disabled: this.disabled,
        connected: this.connected,
        toggleLogin: this.toggleLogin,
        disconnected: this.disconnected
      }
    }
  },
  methods: {
    toggleLogin() {
      this.connected ? this.logout() : this.login()
    },
    async getSdk() {
      const { appId, version, options } = this
      const promise = getSdk({ appId, version, ...options })
      const sdk = await this.catchAsync(promise)
      return sdk
    },
    async login() {
      const promise = login(this.loginOptions)
      const { status } = await this.async(promise)
      if (status === 'connected') {
        this.connected = true
        this.$emit('login', status)
      }
      return promise
    },
    async logout() {
      await this.async(logout())
      this.connected = false
      this.$emit('logout')
      return logout
    },
    async catchAsync(promise) {
      return promise.catch(error => {
        if (error) {
          this.error = error
          console.error(this.error)
        }
      })
    },
    async async(promise) {
      this.working = true
      const handled = await this.catchAsync(promise)
      this.working = false
      return handled
    }
  },
  render() {
    if (this.$scopedSlots.default) {
      return this.$scopedSlots.default(this.scope)
    }
  }
}
