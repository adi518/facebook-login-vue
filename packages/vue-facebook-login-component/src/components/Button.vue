<template>
  <!-- https://developers.facebook.com/docs/facebook-login/userexperience/#buttondesign -->
  <scope
    v-model="model"
    v-bind="$props"
    v-on="scopeListeners"
    @sdk-init="handleSdkInit"
  >
    <button
      slot-scope="scope"
      @click="handleClick"
      class="v-facebook-login"
      :disabled="scope.disabled"
      :style="computedButtonStyle"
    >
      <slot name="before" v-bind="scope"></slot>
      <slot name="loader" v-bind="scope">
        <span class="loader" v-if="scope.working" :style="loaderStyle"></span>
      </slot>
      <slot name="logo" v-bind="scope" v-if="scope.idle && scope.disconnected">
        <template v-if="useAlternateLogo">
          <v-facebook-logo-alt
            class="logo"
            :class="logoClass"
            :style="logoStyle"
          ></v-facebook-logo-alt>
        </template>
        <template v-else>
          <v-facebook-logo
            class="logo"
            :class="logoClass"
            :style="logoStyle"
          ></v-facebook-logo>
        </template>
      </slot>
      <span :style="textStyle">
        <slot
          name="login"
          v-bind="scope"
          v-if="scope.idle && scope.disconnected"
          >Continue with Facebook</slot
        >
        <slot
          name="logout"
          v-bind="scope"
          v-if="scope.idle && scope.enabled && scope.connected"
          >Logout</slot
        >
        <slot name="working" v-bind="scope" v-if="scope.working"
          >Please wait...</slot
        >
        <slot name="error" v-bind="scope" v-if="scope.error">â›” Error</slot>
      </span>
      <slot name="after" v-bind="scope"></slot>
    </button>
  </scope>
</template>

<script>
import Logo from './Logo'
import Scope from './Scope'
import LogoAlt from './LogoAlt'

export default {
  name: 'v-facebook-login',
  components: {
    Scope,
    VFacebookLogo: Logo,
    VFacebookLogoAlt: LogoAlt
  },
  props: {
    ...Scope.props,
    ...{
      logoClass: {
        type: String
      },
      buttonStyle: {
        type: Object,
        default: () => ({})
      },
      loaderStyle: {
        type: Object,
        default: () => ({})
      },
      logoStyle: {
        type: Object,
        default: () => ({})
      },
      textStyle: {
        type: Object,
        default: () => ({})
      },
      transition: {
        type: Array,
        default: () => []
      },
      useAlternateLogo: {
        type: Boolean,
        default: false
      }
    }
  },
  data() {
    // eslint-disable-next-line no-unused-vars
    const { click, ...scopeListeners } = this.$listeners
    return { scope: {}, scopeListeners }
  },
  computed: {
    model: {
      get() {
        return this.value
      },
      set(value) {
        this.$emit('input', value)
      }
    },
    computedTransition() {
      const transition = ['background-color 0.15s ease-in-out'].concat(
        this.transition
      )
      return {
        transition: [...new Set(transition)].join(', ')
      } // Bootstrap 4 (`.btn`)
    },
    computedButtonStyle() {
      return { ...this.buttonStyle, ...this.computedTransition }
    }
  },
  methods: {
    handleSdkInit({ scope }) {
      this.scope = scope
    },
    handleClick() {
      this.$emit('click')
      this.scope.toggleLogin()
    }
  }
}
</script>

<style lang="scss" scoped>
// http://www.color-hex.com/color/1c284c
// https://iconmonstr.com/facebook-1-svg/
// http://www.color-hex.com/color-palette/185
// https://icons8.com/icon/set/facebook-f/all

$color-white: #ffffff;
$color-nepal: #8b9dc3;
$color-chambray: #3b55a0;

.v-facebook-login {
  margin: 0; // Normalize Flex-box
  display: flex;
  cursor: default; // Normalize IE 11
  min-width: 15rem;
  color: $color-white;
  align-items: center;
  box-sizing: border-box;
  border-radius: 0.25rem;
  justify-content: center;
  padding: 0.5rem 1.275rem;
  border: 1px solid rgba($color-white, 0.05);
  background-color: lighten($color-chambray, 1%);

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  &[disabled] {
    opacity: 0.75;
    cursor: not-allowed;
    background-color: darken($color-chambray, 10%);

    &:hover {
      background-color: darken($color-chambray, 10%);
    }
  }

  &:hover {
    background-color: lighten($color-chambray, 5%);
  }

  &:focus {
    outline-width: 0;
    box-shadow: 0 0 0 1px rgba($color-nepal, 0.5);
  }
}

.logo {
  height: auto;
  margin-right: 1.5ch;
}

.loader {
  display: block;
  border-radius: 50%;
  border-style: solid;
  border-width: 0.1rem;
  border-top-color: $color-nepal;
  border-right-color: rgba($color-white, 1);
  border-left-color: rgba($color-white, 1);
  border-bottom-color: rgba($color-white, 1);
  animation: spin 2s linear infinite;
}

.loader {
  height: 1.5rem;
  margin-right: 1ch;
}

.logo,
.loader {
  opacity: 0.9;
  width: 1.5rem;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
</style>
