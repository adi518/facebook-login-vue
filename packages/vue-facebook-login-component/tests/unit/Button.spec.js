import flushPromises from 'flush-promises'
import { mount } from '@vue/test-utils'
import Button from '@/components/Button'
import Scope from '@/components/Scope'

beforeEach(() => {
  jest.spyOn(console, 'error').mockImplementation(() => {})
})

jest.mock('@/helpers', () => ({
  __esModule: true,
  removeSdkScript: async () => undefined,
  getSdk: jest.fn().mockResolvedValue({}),
  login: jest.fn().mockResolvedValue({ status: 'connected' }),
  logout: jest.fn().mockResolvedValue(),
  getLoginStatus: jest
    .fn()
    .mockResolvedValueOnce({ status: 'unknown' })
    .mockResolvedValueOnce({ status: 'unknown' })
    .mockResolvedValueOnce({ status: 'connected' })
    .mockResolvedValueOnce({ status: 'unknown' })
    .mockResolvedValueOnce({ status: 'unknown' })
    .mockResolvedValueOnce({ status: 'unknown' })
    .mockResolvedValueOnce({ status: 'unknown' })
    .mockResolvedValueOnce({ status: 'connected' })
    .mockResolvedValueOnce({ status: 'connected' })
}))

describe('Button', () => {
  test('initial render (disabled)', async () => {
    const wrapper = mount(Button)
    await flushPromises()
    expect(wrapper.attributes('disabled')).toBe('disabled')
    expect(wrapper.element).toMatchSnapshot()
  })

  test('initial render (disconnected)', async () => {
    const wrapper = mount(Button, {
      propsData: { appId: '966242223397117' }
    })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('initial render (connected)', async () => {
    const wrapper = mount(Button, {
      propsData: { appId: '966242223397117' }
    })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('login and click event', async () => {
    const onClick = jest.fn()
    const wrapper = mount(Button, {
      propsData: { appId: '966242223397117' },
      listeners: { click: onClick }
    })
    wrapper.get('button').trigger('click')
    await wrapper.vm.$nextTick()
    await flushPromises()
    // expect(onClick).toHaveBeenCalled()
    // expect(wrapper.emitted().input.slice(-1)[0]).toEqual([
    //   {
    //     connected: true,
    //     disabled: false,
    //     disconnected: false,
    //     enabled: true,
    //     error: null,
    //     idle: true,
    //     working: false
    //   }
    // ])
  })

  test('props', async () => {
    const propsData = {
      appId: '966242223397117',
      version: 'mock-version',
      value: { connected: false },
      options: { option1: 1, option2: 2, option3: 3 },
      textStyle: { textStyle1: 1, textStyle2: 2, textStyle3: 3 },
      logoStyle: { logoStyle1: 1, logoStyle2: 2, logoStyle3: 3 },
      loaderStyle: { loaderStyle1: 1, loaderStyle2: 2, loaderStyle3: 3 },
      buttonStyle: { buttonStyle1: 1, buttonStyle2: 2, buttonStyle3: 3 },
      loginOptions: { loginOption1: 1, loginOption2: 2, loginOption3: 3 },
      useAlternateLogo: true,
      transition: [
        'background-color 0.15s ease-in-out',
        'padding 0.15s ease-in-out'
      ]
    }
    const wrapper = mount(Button, { propsData })
    expect(wrapper.vm.$props).toEqual(propsData)
  })

  test('slots', async () => {
    const wrapper = mount(Button, {
      propsData: { appId: '966242223397117' },
      slots: {
        logo: '<div class="logo"></div>',
        after: '<div class="after"></div>',
        before: '<div class="before"></div>',
        login: '<div class="login"></div>',
        error: '<div class="error"></div>',
        working: '<div class="working"></div>'
      }
    })
    await flushPromises()
    const scopeInstance = wrapper.get(Scope).vm
    scopeInstance.error = new Error()
    scopeInstance.$forceUpdate()
    await scopeInstance.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('scoped-slots', async () => {
    const wrapper = mount(Button, {
      propsData: { appId: '966242223397117' },
      scopedSlots: {
        logo: '<div class="logo" slot-scope="scope">{{scope}}</div>',
        after: '<div class="after" slot-scope="scope">{{scope}}</div>',
        before: '<div class="before" slot-scope="scope">{{scope}}</div>',
        login: '<div class="login" slot-scope="scope">{{scope}}</div>',
        error: '<div class="error" slot-scope="scope">{{scope}}</div>',
        working: '<div class="working" slot-scope="scope">{{scope}}</div>'
      }
    })
    await flushPromises()
    const scopeInstance = wrapper.get(Scope).vm
    scopeInstance.error = new Error()
    scopeInstance.$forceUpdate()
    await scopeInstance.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('logout slot', async () => {
    const wrapper = mount(Button, {
      propsData: { appId: '966242223397117' },
      slots: { logout: '<div class="logout"></div>' }
    })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('logout scoped-slot', async () => {
    const wrapper = mount(Button, {
      propsData: { appId: '966242223397117' },
      scopedSlots: {
        logout: '<div class="logout" slot-scope="scope">{{scope}}</div>'
      }
    })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
  })
})
