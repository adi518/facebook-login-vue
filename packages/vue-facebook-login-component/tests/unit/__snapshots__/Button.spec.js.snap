// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button initial render (connected) 1`] = `
<button
  buttonstyle="[object Object]"
  class="v-facebook-login"
  loaderstyle="[object Object]"
  logostyle="[object Object]"
  style="transition: background-color 0.15s ease-in-out;"
  textstyle="[object Object]"
  transition=""
>
   
  <!---->
   
  <!---->
   
  <span>
    <!---->
     Logout 
    <!---->
     
    <!---->
  </span>
   
</button>
`;

exports[`Button initial render (disabled) 1`] = `
<button
  buttonstyle="[object Object]"
  class="v-facebook-login"
  disabled="disabled"
  loaderstyle="[object Object]"
  logostyle="[object Object]"
  style="transition: background-color 0.15s ease-in-out;"
  textstyle="[object Object]"
  transition=""
>
   
  <!---->
   
  <svg
    class="logo"
    fill="currentColor"
    height="24"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M22.675 0h-21.35c-.732 0-1.325.593-1.325 1.325v21.351c0 .731.593 1.324 1.325 1.324h11.495v-9.294h-3.128v-3.622h3.128v-2.671c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.795.143v3.24l-1.918.001c-1.504 0-1.795.715-1.795 1.763v2.313h3.587l-.467 3.622h-3.12v9.293h6.116c.73 0 1.323-.593 1.323-1.325v-21.35c0-.732-.593-1.325-1.325-1.325z"
    />
  </svg>
   
  <span>
    Continue with Facebook 
    <!---->
     
    <!---->
     
    <!---->
  </span>
   
</button>
`;

exports[`Button initial render (disconnected) 1`] = `
<button
  buttonstyle="[object Object]"
  class="v-facebook-login"
  loaderstyle="[object Object]"
  logostyle="[object Object]"
  style="transition: background-color 0.15s ease-in-out;"
  textstyle="[object Object]"
  transition=""
>
   
  <!---->
   
  <svg
    class="logo"
    fill="currentColor"
    height="24"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M22.675 0h-21.35c-.732 0-1.325.593-1.325 1.325v21.351c0 .731.593 1.324 1.325 1.324h11.495v-9.294h-3.128v-3.622h3.128v-2.671c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.795.143v3.24l-1.918.001c-1.504 0-1.795.715-1.795 1.763v2.313h3.587l-.467 3.622h-3.12v9.293h6.116c.73 0 1.323-.593 1.323-1.325v-21.35c0-.732-.593-1.325-1.325-1.325z"
    />
  </svg>
   
  <span>
    Continue with Facebook 
    <!---->
     
    <!---->
     
    <!---->
  </span>
   
</button>
`;

exports[`Button logout scoped-slot 1`] = `
<button
  buttonstyle="[object Object]"
  class="v-facebook-login"
  loaderstyle="[object Object]"
  logostyle="[object Object]"
  style="transition: background-color 0.15s ease-in-out;"
  textstyle="[object Object]"
  transition=""
>
   
  <!---->
   
  <!---->
   
  <span>
    <!---->
     
    <div
      class="logout"
    >
      {
  "idle": true,
  "error": null,
  "working": false,
  "enabled": true,
  "disabled": false,
  "connected": true,
  "disconnected": false
}
    </div>
     
    <!---->
     
    <!---->
  </span>
   
</button>
`;

exports[`Button logout slot 1`] = `
<button
  buttonstyle="[object Object]"
  class="v-facebook-login"
  loaderstyle="[object Object]"
  logostyle="[object Object]"
  style="transition: background-color 0.15s ease-in-out;"
  textstyle="[object Object]"
  transition=""
>
   
  <!---->
   
  <!---->
   
  <span>
    <!---->
     
    <div
      class="logout"
    />
     
    <!---->
     
    <!---->
  </span>
   
</button>
`;

exports[`Button scoped-slots 1`] = `
<button
  buttonstyle="[object Object]"
  class="v-facebook-login"
  disabled="disabled"
  loaderstyle="[object Object]"
  logostyle="[object Object]"
  style="transition: background-color 0.15s ease-in-out;"
  textstyle="[object Object]"
  transition=""
>
  <div
    class="before"
  >
    {
  "idle": true,
  "error": {},
  "working": false,
  "enabled": false,
  "disabled": true,
  "connected": false,
  "disconnected": true
}
  </div>
   
  <!---->
   
  <div
    class="logo"
  >
    {
  "idle": true,
  "error": {},
  "working": false,
  "enabled": false,
  "disabled": true,
  "connected": false,
  "disconnected": true
}
  </div>
   
  <span>
    <div
      class="login"
    >
      {
  "idle": true,
  "error": {},
  "working": false,
  "enabled": false,
  "disabled": true,
  "connected": false,
  "disconnected": true
}
    </div>
     
    <!---->
     
    <!---->
     
    <div
      class="error"
    >
      {
  "idle": true,
  "error": {},
  "working": false,
  "enabled": false,
  "disabled": true,
  "connected": false,
  "disconnected": true
}
    </div>
  </span>
   
  <div
    class="after"
  >
    {
  "idle": true,
  "error": {},
  "working": false,
  "enabled": false,
  "disabled": true,
  "connected": false,
  "disconnected": true
}
  </div>
</button>
`;

exports[`Button slots 1`] = `
<button
  buttonstyle="[object Object]"
  class="v-facebook-login"
  disabled="disabled"
  loaderstyle="[object Object]"
  logostyle="[object Object]"
  style="transition: background-color 0.15s ease-in-out;"
  textstyle="[object Object]"
  transition=""
>
  <div
    class="before"
  />
   
  <!---->
   
  <div
    class="logo"
  />
   
  <span>
    <div
      class="login"
    />
     
    <!---->
     
    <!---->
     
    <div
      class="error"
    />
  </span>
   
  <div
    class="after"
  />
</button>
`;
